system/Helpers/form_helper.php

```diff
@@ -0,0 +1,859 @@
+     * @return	string
+     */
+    function form_dropdown($data = '', $options = [], $selected = [], $extra = ''): string
+    {
+        $defaults = [];
+        if (is_array($data))
+        {
+            if (isset($data['selected']))
+            {
+                $selected = $data['selected'];
+                unset($data['selected']); // select tags don't have a selected attribute
+            }
+            if (isset($data['options']))
+            {
+                $options = $data['options'];
+                unset($data['options']); // select tags don't use an options attribute
+            }
+        }
+        else
+        {
+            $defaults = ['name' => $data];
+        }
+        is_array($selected) OR $selected = [$selected];
+        is_array($options) OR $options = [$options];
+        // If no selected state was submitted we will attempt to set it automatically
+        if (empty($selected))
+        {
+            if (is_array($data))
+            {
+                if (isset($data['name'], $_POST[$data['name']]))
+                {
+                    $selected = [$_POST[$data['name']]];
+                }
+            }
+            elseif (isset($_POST[$data]))
+            {
+                $selected = [$_POST[$data]];
+            }
+        }
+        $extra = _attributes_to_string($extra);
+        $multiple = (count($selected) > 1 && stripos($extra, 'multiple') === false) ? ' multiple="multiple"' : '';
+        $form = '<select '.rtrim(_parse_form_attributes($data, $defaults)).$extra.$multiple.">\n";
+        foreach ($options as $key => $val)
+        {
+            $key = (string) $key;
+            if (is_array($val))
+            {
+                if (empty($val))
+                {
+                    continue;
+                }
+                $form .= '<optgroup label="'.$key."\">\n";
+                foreach ($val as $optgroup_key => $optgroup_val)
+                {
+                    $sel = in_array($optgroup_key, $selected) ? ' selected="selected"' : '';
+                    $form .= '<option value="'.htmlspecialchars($optgroup_key).'"'.$sel.'>'
+                            .(string) $optgroup_val."</option>\n";
+                }
+                $form .= "</optgroup>\n";
+            }
+            else
+            {
+                $form .= '<option value="'.htmlspecialchars($key).'"'
+                        .(in_array($key, $selected) ? ' selected="selected"' : '').'>'
+                                .(string) $val."</option>\n";
+            }
+        }
+        return $form."</select>\n";
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_checkbox'))
+{
+    /**
+     * Checkbox Field
+     *
+     * @param	mixed
+     * @param	string
+     * @param	bool
+     * @param	mixed
+     * @return	string
+     */
+    function form_checkbox($data = '', string $value = '', bool $checked = false, $extra = ''): string
```

Comment:
Missing a couple of type hints. Please verify that all methods use type hints where possible.